language/libra-tools/writeset-transaction-generator/src/writeset_builder.rs

```diff
@@ -0,0 +1,180 @@
+            libra_root_address(),
+            "LibraConfig",
+            "disable_reconfiguration",
+            vec![],
+            vec![Value::transaction_argument_signer_reference(
+                libra_root_address(),
+            )],
+        )
+    }
+
+    fn enable_reconfiguration(&mut self) {
+        self.exec_func(
+            libra_root_address(),
+            "LibraConfig",
+            "enable_reconfiguration",
+            vec![],
+            vec![Value::transaction_argument_signer_reference(
+                libra_root_address(),
+            )],
+        )
+    }
+    pub fn set_libra_version(&mut self, version: Version) {
+        self.exec_func(
+            libra_root_address(),
+            "LibraVersion",
+            "set",
+            vec![],
+            vec![
+                Value::transaction_argument_signer_reference(libra_root_address()),
+                Value::u64(version),
+            ],
+        )
+    }
+}
+
+/// Convert the transaction arguments into Move values.
+fn convert_txn_args(args: &[TransactionArgument]) -> Vec<Value> {
+    args.iter()
+        .map(|arg| match arg {
+            TransactionArgument::U8(i) => Value::u8(*i),
+            TransactionArgument::U64(i) => Value::u64(*i),
+            TransactionArgument::U128(i) => Value::u128(*i),
+            TransactionArgument::Address(a) => Value::address(*a),
+            TransactionArgument::Bool(b) => Value::bool(*b),
+            TransactionArgument::U8Vector(v) => Value::vector_u8(v.clone()),
+        })
+        .collect()
+}
+
+fn move_module_changes(modules: &[CompiledModule]) -> MoveChanges {
+    let mut shadow_changeset = MoveChanges::new();
+    for module in modules.iter() {
+        let id = module.self_id();
+        let mut module_bytes = vec![];
+        module.serialize(&mut module_bytes).unwrap();
+        shadow_changeset.publish_or_overwrite_module(id, module_bytes);
+    }
+    shadow_changeset
+}
+
+pub fn build_changeset<S: StateView, F>(
+    state_view: &S,
+    procedure: F,
+    modules: &[CompiledModule],
+) -> ChangeSet
+where
+    F: FnOnce(&mut GenesisSession<DeltaStorage<RemoteStorage<S>>>),
+{
+    let move_vm = MoveVM::new();
+    let move_changes = move_module_changes(modules);
+    let mut effect = {
+        let state_view_storage = RemoteStorage::new(state_view);
+        let exec_storage = DeltaStorage::new(&state_view_storage, &move_changes);
+        let mut session = GenesisSession(move_vm.new_session(&exec_storage));
+        session.disable_reconfiguration();
+        procedure(&mut session);
+        session.enable_reconfiguration();
+        session
+            .0
+            .finish()
+            .map_err(|err| format_err!("Unexpected VM Error: {:?}", err))
+            .unwrap()
+    };
+
+    for module in modules {
+        let mut module_bytes = vec![];
+        module.serialize(&mut module_bytes).unwrap();
+        // TODO: Check compatibility between old and new modules.
```

Comment:
This should be easier to do once https://github.com/libra/libra/pull/6753 lands (see new `compatibility.rs` file). But definitely seems like we should add this sanity check before landing.